lib/src/lib.rs

```diff
@@ -1007,171 +459,164 @@ pub struct ColourCmd {
 }
 
 /// [Line Type 0](https://www.ldraw.org/article/218.html#lt0) comment.
-#[derive(Debug, PartialEq)]
+#[derive(Debug, PartialEq, Clone)]
 pub struct CommentCmd {
     /// Comment content, excluding the command identififer `0` and the optional comment marker `//`.
     pub text: String,
 }
 
+impl CommentCmd {
+    pub fn new(text: &str) -> CommentCmd {
+        CommentCmd {
+            text: text.to_string(),
+        }
+    }
+}
+
+/// [Line Type 0](https://www.ldraw.org/article/218.html#lt0) FILE start.
+/// [MPD Extension[(https://www.ldraw.org/article/47.html)
+#[derive(Debug, PartialEq, Clone)]
+pub struct FileCmd {
+    /// The filename for this file.
+    pub file: String,
+}
+
+/// [Line Type 0](https://www.ldraw.org/article/218.html#lt0) DATA.
+/// [MPD Extension[(https://www.ldraw.org/article/47.html)
+#[derive(Debug, PartialEq, Clone)]
+pub struct DataCmd {
+    /// The filename for this data file.
+    pub file: String,
+}
+
+/// [Line Type 0](https://www.ldraw.org/article/218.html#lt0) base64 data chunk.
+/// [MPD Extension[(https://www.ldraw.org/article/47.html)
+#[derive(Debug, PartialEq, Clone)]
+pub struct Base64DataCmd {
+    /// The decoded base64 data chunk.
+    pub data: Vec<u8>,
+}
+
 /// Single LDraw source file loaded and optionally parsed.
-#[derive(Debug, PartialEq)]
+#[derive(Debug, PartialEq, Clone)]
 pub struct SourceFile {
     /// The relative filename of the file as resolved.
     pub filename: String,
-    /// Raw UTF-8 file content (without BOM) loaded from the resolved file. Line ending can be
-    /// Unix style `\n` or Windows style `\r\n`. As a convenience, parsing handles both indifferently,
-    /// so the file can contain a mix of both, although this is not recommended.
-    pub raw_content: Vec<u8>,
     /// LDraw commands parsed from the raw text content of the file.
     pub cmds: Vec<Command>,
 }
 
 /// Collection of [`SourceFile`] accessible from their reference filename.
 #[derive(Debug)]
 pub struct SourceMap {
-    /// Array of source files in the collection.
-    source_files: Vec<SourceFile>,
-
-    /// Map table of source files indices into [`SourceMap::source_files`] from
-    /// their reference filename, as it appears in [`SubFileRefCmd`].
-    filename_map: HashMap<String, usize>,
-}
-
-/// Reference to a single [`SourceFile`] instance in a given [`SourceMap`].
-#[derive(Debug, Copy, Clone, Default, PartialEq, PartialOrd, Eq, Ord, Hash)]
-pub struct SourceFileRef {
-    index: usize,
-    // TODO: ref to SourceMap?
-}
-
-impl CommentCmd {
-    pub fn new(text: &str) -> CommentCmd {
-        CommentCmd {
-            text: text.to_string(),
-        }
-    }
-}
-
-impl SourceFileRef {
-    pub fn get<'a>(&'a self, source_map: &'a SourceMap) -> &'a SourceFile {
-        source_map.get(*self)
-    }
+    /// Map of filenames to source files.
+    // TODO: How to handle case sensitivity like STUD.DAT vs stud.dat?
```

Comment:
That assumes the LDraw parts library uses a consistent casing for filenames. It's worth making a separate issue for this, since I'm not sure if there's a "correct" way of handling it. If the command references "STUD.DAT" but the file system has both "stud.dat" and "Stud.DAT", which file do we pick?

Reply:
Apparently "STUD.DAT" is a weird special case. I've created https://github.com/djeedai/weldr/issues/30 to track this.

Reply:
> That assumes the LDraw parts library uses a consistent casing for filenames.

You mean my suggestion? No it's not, it's working around that:

```rust
struct SourceFileRef(String);

impl SourceFileRef {
  fn new(str: impl Into<String>) -> Self {
    let str = str.into();
    Self(str.to_lowercase())
  }
}

impl Hash for SourceFileRef {
  fn hash<H: Hasher>(&self, state: &mut H) {
    self.0.hash(state);
  }
}

impl PartialEq for SourceFileRef {
  fn eq(&self, other: &Self) -> bool {
    self.0 == other.0
  }
}

impl Eq for SourceFileRef {}
```

> If the command references "STUD.DAT" but the file system has both "stud.dat" and "Stud.DAT", which file do we pick?

There's no solution. The specification says it should be case-insensitive, so if that situation happens the scenario itself is ill-defined and there is no unique solution:
- you could decide to pick one randomly, assuming they're exact duplicates, but that may produce inconsistent results if they're not.
- you could decide to try and find an exact case-sensitive match, assuming the 2 files are different, but that goes against the specification.

I think we should just ignore that case. If the user or calling code put itself in that situation they shouldn't expect us to solve it without their help. So just doing all comparisons case-insensitive should be enough to handle all well-formed usages.

> Apparently "STUD.DAT" is a weird special case. I've created #30 to track this.

The forum you linked has a few posts explicitly saying "please do ignore this, it's undocumented and nobody knows about it except `ldraw.exe`, you should absolutely ignore that rule about STUD.DAT". I agree with their analysis.